### ISSUE TITLE
How to include custom hyper params file in python?

### ISSUE THREAD

**Issue by fridary** [ultralytics/ultralytics#5048]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I need to turn off any image rotating, I assume there is only `fliplr` param that needs to set 0. For that was created `hyp.yaml` file. To include it:
`python train.py --hyp hyp.yaml`

How to include it in python?
`model.train(data='data.yaml', hyp='hyp.yaml')` ?

### Additional

I have not found answer here https://docs.ultralytics.com/modes/train/

**Comment by glenn-jocher** [comment]
@fridary sure, I can certainly guide you on how to include your custom hyperparameters in Python.

The command `model.train(data='data.yaml', hyp='hyp.yaml')` is what you'd typically use to include your custom `hyp.yaml` file while initiating the training process.

In your `hyp.yaml` file, you would set `fliplr: 0` to disable horizontal flipping (image rotation) during augmentation.

Now, to include this in Python, you will need to load your hyperparameters from the `hyp.yaml` file and then pass them to whatever function/method that requires it. Python's built-in `yaml` module can be used to load this file.

For example, consider you have a function `train_yolov8()`. This method would need to have an argument for hyperparameters (let's call it `hyp`). So when you call this function, pass your loaded `hyp.yaml` file to it, like so: `train_yolov8(hyp=my_loaded_hyp)`.

Remember, in the scope of this function/method, `hyp` would be a dictionary with keys and values matching what you have in your `hyp.yaml` file.

It's noteworthy that different models might handle these image augmentations differently, so ensure to check any additional model-specific parameters if needed. I hope this guidance helps in your work with YOLOv8!

**Comment by fridary** [comment]
Thank you for detailed answer, I read many times and absolutely did not understand. What is `train_yolov8`, where it comes from? I simply need to `train()` with custom hyper params, how to do that?
If I execute `model.train(data='data.yaml', hyp='hyp.yaml')`, I will get error `SyntaxError: 'hyper' is not a valid YOLO argument`. How to put custom `hyp.yaml` file in `model.train()`?

```python
from ultralytics import YOLO
model = YOLO("yolov8m.yaml")
model.train(data='data.yaml', hyper='hyper.yaml', batch=8, imgsz=640, epochs=600, verbose=True)
```

```
Traceback (most recent call last):
  File "/home/fridary/python/yolo/train.py", line 21, in <module>
    model.train(data='data.yaml', hyper='hyper.yaml', batch=8, imgsz=640, epochs=600, verbose=True) # classif
  File "/home/fridary/miniconda3/envs/yolo/lib/python3.9/site-packages/ultralytics/engine/model.py", line 336, in train
    self.trainer = (trainer or self.smart_load('trainer'))(overrides=args, _callbacks=self.callbacks)
  File "/home/fridary/miniconda3/envs/yolo/lib/python3.9/site-packages/ultralytics/engine/trainer.py", line 83, in __init__
    self.args = get_cfg(cfg, overrides)
  File "/home/fridary/miniconda3/envs/yolo/lib/python3.9/site-packages/ultralytics/cfg/__init__.py", line 114, in get_cfg
    check_dict_alignment(cfg, overrides)
  File "/home/fridary/miniconda3/envs/yolo/lib/python3.9/site-packages/ultralytics/cfg/__init__.py", line 203, in check_dict_alignment
    raise SyntaxError(string + CLI_HELP_MSG) from e
SyntaxError: 'hyper' is not a valid YOLO argument. Similar arguments are i.e. ['shear=0.0'].

    Arguments received: ['yolo']. Ultralytics 'yolo' commands use the following syntax:

        yolo TASK MODE ARGS

        Where   TASK (optional) is one of ('detect', 'segment', 'classify', 'pose')
                MODE (required) is one of ('train', 'val', 'predict', 'export', 'track', 'benchmark')
                ARGS (optional) are any number of custom 'arg=value' pairs like 'imgsz=320' that override defaults.
                    See all ARGS at https://docs.ultralytics.com/usage/cfg or with 'yolo cfg'

    1. Train a detection model for 10 epochs with an initial learning_rate of 0.01
        yolo train data=coco128.yaml model=yolov8n.pt epochs=10 lr0=0.01

    2. Predict a YouTube video using a pretrained segmentation model at image size 320:
        yolo predict model=yolov8n-seg.pt source='https://youtu.be/Zgi9g1ksQHc' imgsz=320

    3. Val a pretrained detection model at batch-size 1 and image size 640:
        yolo val model=yolov8n.pt data=coco128.yaml batch=1 imgsz=640

    4. Export a YOLOv8n classification model to ONNX format at image size 224 by 128 (no TASK required)
        yolo export model=yolov8n-cls.pt format=onnx imgsz=224,128

    5. Run special commands:
        yolo help
        yolo checks
        yolo version
        yolo settings
        yolo copy-cfg
        yolo cfg

    Docs: https://docs.ultralytics.com
    Community: https://community.ultralytics.com
    GitHub: https://github.com/ultralytics/ultralytics
```

**Comment by RizwanMunawar** [comment]
Hello @fridary! Just to let you know, modifying the `hyper` or `hyp` parameters during training isn't supported. You can find more detailed information on this in the [Ultralytics YOLOv8 training documentation](https://docs.ultralytics.com/modes/train/#arguments).

However, if you do wish to make changes to hyperparameters, you can follow these steps:

1. First, uninstall the `ultralytics` package if you've previously installed it using the command `pip uninstall ultralytics`.
2. Next, clone the `ultralytics` repository with this command: `git clone https://github.com/ultralytics/ultralytics.git`.
3. Navigate to the `ultralytics/cfg/default.yaml` file, search for `fliplr`, modify its value as needed, and then save the file.
4. Open the cloned `ultralytics` folder in your CLI and execute the command `python setup.py install`.

Once this process is complete, the `ultralytics` package will be installed with your specified `fliplr` value, set to `0` for each training. If you wish to modify other parameters, simply repeat steps 3 and 4 as necessary.

Hope this helps! Thanks

**Comment by youngjaean** [comment]
```
from ultralytics import YOLO

# Load a model
model = YOLO("yolov8n.yaml")  # build a new model from scratch
model = YOLO("yolov8n.pt")  # load a pretrained model (recommended for training)

# Use the model
model.train(data="data.yaml", epochs=80,  cfg='custom_hyp.yaml')  

```

this is more than better

**Comment by glenn-jocher** [comment]
@youngjaean hello, thank you for your suggestion. The usage you're suggesting does allow for greater flexibility by using a custom `hyp.yaml` file. 

However, it's important to note that in the current architecture of YOLOv8's Ultralytics implementation, parameters from a `hyp.yaml` file cannot be directly replaced during the execution of the `model.train()` function.

The training hyperparameters are loaded as the model is initialised and are kept constant throughout the model's lifecycle. If you need to alter these hyperparameters, you would need to do so directly in the source code of the specific YOLOv8 version you're using.

This design choice is in line with best practices in machine learning, where hyperparameters are usually set before training begins and kept constant throughout training.

Your suggestion is valid and can be a good area for development in future versions. It encourages further dynamism and flexibility in the model's usability. Your feedback is very much appreciated, and please don't hesitate to let us know if you have any other questions or suggestions!

**Comment by oukohou** [comment]
seems like this hyper parameter is configured in `ultralytics/utils/__init__.py` line 35:
`DEFAULT_CFG_PATH = ROOT / 'cfg/default.yaml'`
maybe modify this before you `from ultralytics import YOLO`?

**Comment by glenn-jocher** [comment]
@oukohou indeed, in the Ultralytics YOLOv8 package, the default hyperparameter configuration is specified in the `DEFAULT_CFG_PATH`. Modifications to this file will alter the hyperparameters used during model training and other processes.

To use a custom hyperparameter file, you could overwrite the `default.yaml` file with your `hyp.yaml` before importing and initializing the `YOLO` class. However, direct manipulation of the default configuration path in the package is not generally recommended as it makes the codebase unpredictable, less maintainable, and it affects the global package behavior which might not be intended.

A more practical and safer approach would be to clone the repository, modify the `default.yaml` within your cloned folder (now a standalone package), and install it locally. This ensures your global environment remains unaffected by the changes, and you'd be running a version of the package with your desired hyperparameters.

In summary, while programmatically altering the `DEFAULT_CFG_PATH` could achieve what you're asking for, it would not be considered a best practice. Instead, the recommended approach is to work with a dedicated local copy of the Ultralytics YOLO package for your custom implementations.

**Comment by moji-programming** [comment]
It's really easy, just several lines of code:

```
cfg_file = "data_hsv.yaml"
cfg = {}
with open(cfg_file, "r") as f:
    cfg = yaml.load(f, Loader=yaml.FullLoader)
   for item in ["path", "train", "val", "test", "names", ]:
        cfg.pop(item, None)
model = YOLO("yolov8n.pth")
model.train(data=cfg_file, **cfg)
```

**Comment by glenn-jocher** [comment]
Thank you for sharing your approach! For YOLO11 users, we recommend using the native hyperparameter handling through separate `hyp.yaml` files as shown in the [Training Arguments documentation](https://docs.ultralytics.com/modes/train/#arguments). Your method of modifying the config dictionary directly might bypass some internal validation checks. A more supported approach would be:

```python
from ultralytics import YOLO
model = YOLO("yolov8n.pt")  # .pt format recommended
model.train(data="data.yaml", hyp="hyp.yaml")  # keep data/hyp separate
```

This maintains the intended configuration structure while allowing custom hyperparameters.
